modules/standard/Heap.chpl

```diff
@@ -0,0 +1,432 @@
+        var greaterChild = pos*2;
+        if (greaterChild >= _data.size) then break; // reach leaf node, break
+        if (greaterChild + 1 < _data.size) {
+          // if the right child node exists
+          if (_greater(_data[greaterChild+1],_data[greaterChild])) {
+            // if the right child is greater, then update the greaterChild
+            greaterChild += 1;
+          }
+        }
+        // if the greaterChild's value is greater than current node, then swap and continue
+        if (_greater(_data[greaterChild],_data[pos])) {
+          _data[greaterChild] <=> _data[pos];
+          pos = greaterChild;
+        }
+        else break;
+      }
+    }
+
+    /*
+      Push an element into the heap
+
+      :arg element: The element that will be pushed
+      :type element: `eltType`
+    */
+    proc push(in element: eltType)
+    lifetime this < element {
+      _enter();
+      _data.append(element);
+      _heapify_up(_data.size-1);
+      _leave();
+    }
+
+    /*
+      Push elements of a list into the heap
+
+      :arg x: The list of which elements is to push.
+      :type x: `list(eltType)`
+    */
+    proc push(ref x: list(eltType)) {
+      for e in x do
+        push(e);
+    }
+
+    /*
+      Push elements in an array into a heap.
+
+      :arg x: The array of which elements is to push.
+      :type x: `[?d] eltType`
+    */
+    proc push(ref x:[?d] eltType) {
+      for e in x do
+        push(e);
+    }
+
+    /*
+      Pop an element and return it.
+
+      :return: the top element
+      :rtype: eltType
+    */
+    proc pop(): eltType {
+      _enter();
+      if (boundsChecking && isEmpty()) {
+        boundsCheckHalt("Called \"heap.pop\" on an empty heap.");
+      }
+      _data(0) <=> _data(_data.size-1);
+      var ret = _data.pop();
+      _heapify_down(0);
+      _leave();
+      return ret;
+    }
+
+    /*
+      Iterate over the elements of this heap
+
+        .. note::
+          Elements in the heap is not sorted
+
+    */
+    iter these() ref {
+      for e in _data {
+        yield e;
+      }
+    }
+
+    proc const toArray(): [] eltType {
+      if !isCopyableType(eltType) then
+        compilerError("toArray() method is not avaliable on a 'heap'",
+                      " with elements of a type that can't be copied, here: ",
+                      eltType: string);
+      var l: [0..#size] eltType;
+      for i in 0..#size {
+        l[i] = _data[i];
+      }
+      return l;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing the
+      elements contained in this heap. The heap will be empty.
+
+      :return: A new DefaultRectangular array.
+    */
+    proc consume(): [] eltType {
+      var l: [0..#size] eltType;
+      for i in 0..#size {
+        l[i] = pop();
+      }
+      return l;
+    }
+
+    /*
+      Write the contents of this list to a channel.
+
+      :arg ch: A channel to write to.
+    */
+    proc readWriteThis(ch: channel) throws {
```

Comment:
I wrote it assuming that everything it takes is produced by `writeln(someHeap)`. I think it can be useful for restoring states when debugging. Can you describe the edge case more? Is it serious enough to remove the functionality of reading?

Reply:
I'm still hoping @dlongnecke-cray will weigh in, as I think `Heap` and `List` should do the same thing, but I'm having a hard time seeing that what `List` is doing makes sense.

Reply:
Hi, sorry that I haven't responded sooner. I've been following most of the discussion in this PR, though!

I think it would be fine to change this `readWriteThis` function to `writeThis`. Additionally we could go ahead and open an issue about reading containers of string elements (or of elements containing strings), specifically about how we should demarcate such elements (when they can contain a comma themselves).

You could tag `list` in that issue or open a separate issue about changing `list.readWriteThis` to `list.writeThis`.